pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,377 @@
+	}
+
+	glog.V(3).Infof("volume create response : %+v", volumeCreateResp)
+	err = cs.client.VolumeStart(volumeName, true)
+	if err != nil {
+		//TODO : delete the volume
+		glog.Errorf("failed to start volume:%v", err)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to start volume %s", err.Error()))
+	}
+
+	glusterServer, bkpServers, err = cs.getClusterNodes()
+
+	if err != nil {
+                //TODO: delete the volume
+		glog.Errorf("failed to fetch details of cluster nodes: %v", err)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("error in fecthing peer details %s", err.Error()))
+	}
+
+	resp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			Id:            volumeName,
+			CapacityBytes: int64(volSizeBytes),
+			Attributes: map[string]string{
+				"glustervol":        volumeName,
+				"glusterserver":     glusterServer,
+				"glusterbkpservers": strings.Join(bkpServers, ":"),
+			},
+		},
+	}
+
+	glog.V(4).Infof("CSI Volume response: %+v", resp)
+	return resp, nil
+}
+
+func (cs *ControllerServer) checkExistingVolumes(volumeListResp *api.VolumeListResp, volumeName string, volSizeBytes int) (string, []string, error) {
+
+	var tspServers []string
+	var mountServer string
+	var err error
+
+	if volumeListResp == nil {
+		glog.Errorf("Provided volume list response is nil")
+		return "", nil, fmt.Errorf("volume list response is nil")
+	}
+
+	for _, vol := range *volumeListResp {
+		if vol.Name == volumeName {
+			// Do the owner validation
+			if glusterAnnVal, found := vol.Metadata[glusterDescAnn]; found {
+				if glusterAnnVal != glusterDescAnnValue {
+					return "", nil, status.Errorf(codes.Internal, "volume %s (%s) is not owned by Gluster CSI driver",
+						vol.Name, vol.Metadata)
+				}
+			} else {
+				return "", nil, status.Errorf(codes.Internal, "volume %s (%s) is not owned by Gluster CSI driver",
+					vol.Name, vol.Metadata)
+			}
+			vsResp, e := cs.client.VolumeStatus(vol.ID.String())
+			if e != nil {
+				return "", nil, status.Errorf(codes.Internal, "failed to get volume status %s", e.Error())
+			}
+			if int(vsResp.Size.Capacity) != volSizeBytes {
+				return "", nil, status.Error(codes.AlreadyExists, fmt.Sprintf("volume already exits with different size: %d", vsResp.Size.Capacity))
+			}
+
+			glog.Info("Requested volume (%s) already exists in the storage pool", vol.Name)
+			mountServer, tspServers, err = cs.getClusterNodes()
+
+			if err != nil {
+				return "", nil, status.Error(codes.Internal, fmt.Sprintf("error in fetching backup/peer server details %s", err.Error()))
+			}
+
+		}
+	}
+
+	return mountServer, tspServers, nil
+
+}
+func (cs *ControllerServer) getClusterNodes() (string, []string, error) {
+
+	peers, err := cs.client.Peers()
+	if err != nil {
+		return "", nil, err
+	}
+	glusterServer := ""
+	bkpservers := []string{}
+
+	for i, p := range peers {
+		if i == 0 {
+			for _, a := range p.PeerAddresses {
```

Comment:
This should be `ClientAddresses` (same as endpoint)`

Reply:
gd2 runs as a privileged container. but as per default glusterd2 configuration, glusterd2 will listen on all IP addresses, which will have `localhost`,`docker ip ` and other `host ips`  are attached to the container.

we need IP address to mount a volume to app pods.
`PeerAddresses` is having host IP address, which can be used for a mount point.

do we have any use case in running gd2 container on multiple IP?

Reply:
Ideally GD2 should have an API which return the brick servers or backup servers or gluster peer addresses. It will make life a lot easier for integrations. @Madhu-1 @prashanthpai can any one open an issue in GD2 for the same ?

Reply:
I'll open an issue for the same.

The backup servers are slightly less relevant as the glusterfs client (fuse,libgfapi) will send a get this list in the response to `GETSPEC` that happens during mount.

Refer: https://github.com/gluster/glusterd2/issues/382

Reply:
> default glusterd2 configuration, glusterd2 will listen on all IP addresses

The default behaviour will stay. It's configurable though, via the config file or a command line argument and via ENV variables that can be used in containers.

`PeerAddress` should only be used for peer operations - addand remove peer. PeerAddress may go away in future.